### PR TITLE
[reconfigurator] Decommission sleds

### PR DESCRIPTION
**Pull request by jgallagher**
This builds on (and is currently pointed at) #5663, and is on the path to fixing #5625, although there is still more work to do there. (Should be small, and I'll start on it while this is in review to make sure it really fixes it.) I don't plan on landing this before R8 is out the door, but wanted to go ahead and open it up to review.

Much of this is fallout from our discussions about what it means to decommission a sled. Major changes are:

* `SledFilter::All` has been renamed to `SledFilter::Commissioned`, and no longer returns sleds that are in `SledState::Decommissioned`.
* The blueprint planner will now update the desired sled state to `Decommissioned` for sleds which satisfy our conditions. (See `do_plan_decommission()`.)
* The blueprint planner will carry forward the Omicron zones of decommissioned sleds to child blueprints. Pruning these is #5552.
* The blueprint planner will _not_ carry forward a desired sled state of `Decommissioned` once the inputs report that the sled has already been decommissioned.
* The blueprint executor will decommission sleds that the planner said to.
* Decommissioning a sled implicitly decommissions all its disks. (This matches what happens with sled expungement, and should not interfere with region replacement, which keys off of policy, not state.)

### CODE REVIEW

File: dev-tools/omdb/src/bin/omdb/db.rs [hunk 1]
```diff
@@ -1456,7 +1456,7 @@ async fn cmd_db_sleds(
         Some(filter) => filter,
         None => {
             eprintln!("note: listing all sleds (use -F to filter, e.g. -F in-service)");
```
Comment (sunshowers) [inline dev-tools/omdb/src/bin/omdb/db.rs:1458]: note likely needs to be changed

**Review by sunshowers — COMMENTED** [review]
Quick review, but looks reasonable to me!

File: nexus/db-queries/src/db/datastore/sled.rs [hunk 2]
```diff
@@ -1176,6 +1235,29 @@ mod test {
...
+
+        // Observe that the disk state is now decommissioned
+        assert_eq!(
+            PhysicalDiskState::Decommissioned,
+            lookup_physical_disk(&datastore, disk1.id()).await.disk_state
+        );
+        assert_eq!(
+            PhysicalDiskState::Decommissioned,
+            lookup_physical_disk(&datastore, disk2.id()).await.disk_state
+        );
```
Comment (smklein) [inline nexus/db-queries/src/db/datastore/sled.rs:1259]: Thanks for both fixing and testing this

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 3]
```diff
@@ -1036,36 +1082,32 @@ impl<'a> BlueprintZonesBuilder<'a> {
 
     /// Produces an owned map of zones for the requested sleds
     pub fn into_zones_map(
-        mut self,
-        sled_ids: impl Iterator<Item = SledUuid>,
+        self,
+        commissioned_sled_ids: impl Iterator<Item = SledUuid>,
```
Comment (smklein) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:1086]: Nitpick: There really isn't any part of this argument that validates that the state of these sled IDs as "commissioned", right?

Should this just be called "added_sleds"?

**Review by smklein — APPROVED** [review]

**Review by andrewjstone — APPROVED** [review]
Looks great. Mostly nits from me.

File: nexus/db-queries/src/db/datastore/sled.rs [hunk 4]
```diff
@@ -533,85 +534,142 @@ impl DataStore {
...
+                            // To differentiate between the two, check that the
+                            // new state is the same as the old state, and the
+                            // found policy is valid.
+                            if result.found.state() == new_sled_state
+                                && valid_old_policies
+                                    .contains(&result.found.policy())
+                            {
+                                result.found.state()
+                            } else {
+                                return Err(err.bail(
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/db-queries/src/db/datastore/sled.rs:622]: I believe so! I see this pattern in other transactions, and `err.bail`'s comments say it sets `err` to the value we pass here but then returns `DieselError::RollbackTransaction`.

File: nexus/reconfigurator/execution/src/sled_state.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        Err(errors)
+    }
+}
+
+async fn decommission_one_sled(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    sled_id: SledUuid,
+) -> anyhow::Result<()> {
+    let (authz_sled,) = LookupPath::new(opctx, datastore)
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/execution/src/sled_state.rs:43]: `DataStore::sled_set_state_to_decommissioned()` was written in such a way as to require that - it wants an `authz_sled`, so we have to do the lookup to call it at all. I'm not sure why that is, though - @sunshowers do you remember?

File: nexus/reconfigurator/planning/src/planner.rs [hunk 6]
```diff
@@ -100,6 +168,21 @@ impl<'a> Planner<'a> {
             self.blueprint.expunge_all_zones_for_sled(sled_id, reason)?;
         }
 
+        // Check for any decommissioned sleds (i.e., sleds for which our
+        // blueprint has zones, but are not in the input sled list). Any zones
+        // for decommissioned sleds _should_ already be expunged; ensure that is
+        // the case.
+        for sled_id in self.blueprint.sled_ids_with_zones() {
+            if !commissioned_sled_ids.contains(&sled_id) {
+                self.blueprint.expunge_all_zones_for_sled(
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/planner.rs]: Agreed; I changed this to a hard error in 1e629df40

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs (hunk 3 above)
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/builder.rs]: Yeah, that does seem better. Renamed and expanded the doc comment in 64e866024

File: dev-tools/omdb/src/bin/omdb/db.rs (hunk 1 above)
Comment (jgallagher) [reply dev-tools/omdb/src/bin/omdb/db.rs]: Changed the wording here to note it's "all _commissioned_ sleds". The followon PR to this (#5733) adds a `SledFilter::Decommissioned` that allows us to see the sleds that are left out by default here.

**Pull request merged**